drivers: adc: adc_sam: Introduce Atmel SAM ADC driver

pdgendt opened the pull request:
Introduce Atmel SAM ADC driver, and add it to the Atmel SAM4S SoC. 
All of the timings are currently set for the controller instead of individual channels, so only `ADC_ACQ_TIME_DEFAULT` is supported.

Tested on custom board.

@nandojve I will probably need to add this to the `sam4s_xplained` board to take it along in CI. Any suggestions to make it testable?

pdgendt committed:
drivers: adc: adc_sam: Introduce Atmel SAM ADC driver

This commit adds support for Atmel SAM ADC driver with up to
16 channels.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
dts: arm: atmel: Add ADC support to Atmel SAM4S

Add ADC device tree entry for the Atmel SAM4S SoC.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

nandojve commented:
> Introduce Atmel SAM ADC driver, and add it to the Atmel SAM4S SoC. All of the timings are currently set for the controller instead of individual channels, so only `ADC_ACQ_TIME_DEFAULT` is supported.
> 
> Tested on custom board.
> 
> @nandojve I will probably need to add this to the `sam4s_xplained` board to take it along in CI. Any suggestions to make it testable?

Hi @pdgendt,

You can make PB0 and PB1 available as ADC inputs since this is instructed on the SAM4S_Xplained schematic.
About CI, it will be good if we can get both sample and tests running [adc sample](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/adc) + [adc tests](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/adc).

pdgendt committed:
boards: arm: sam4s_xplained: Enable ADC support

Enable ADC entry in device tree and pinctrl (PB0/PB1).
Update doc/yaml files to include ADC info.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
tests: drivers: adc: adc_api: Add sam4s_xplained board

Define ADC channel configuration for the SAM4S Xplained internal
temperature sensor.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
samples: drivers: adc: Add sam4s_xplained overlay

Add a board overlay for sam4s_xplained. Channel 5 (PB1) is enabled as
well as the internal temperature sensor.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
@nandojve I've updated the boards/samples/tests

nandojve commented:
> @nandojve I've updated the boards/samples/tests

Hi @pdgendt ,

This still as a draft or you want to have it on v3.3?

pdgendt commented:
> This still as a draft or you want to have it on v3.3?

Hi @nandojve, I'll clean it up to get it in v3.3.

pdgendt commented:
Hi @nandojve, would you be able to test using the `sam4s_xplained` board?

MaureenHelm commented:
@nandojve ping

nandojve commented:
Hi @pdgendt ,

I configured sample to run as single end instead diff on channel 5 (PB1) and I didn't change temperature configuration.
It seems that there are something not proper with the driver. The read values oscillates and channel 5 not read correct.

The same HW configuration using ASF->SAM->Drivers->ADC sample show different behavior.

```diff
+++ b/samples/drivers/adc/boards/sam4s_xplained.overlay
@@ -9,7 +9,7 @@
 / {
        zephyr,user {
                /* adjust channel number according to pinmux in board.dts */
-               io-channels = <&adc0 2>, <&adc0 15>;
+               io-channels = <&adc0 5>, <&adc0 15>;
        };
 };
 
@@ -17,14 +17,13 @@
        #address-cells = <1>;
        #size-cells = <0>;
 
-       /* External ADC(+)/ADC(-) differential inputs */
-       channel@2 {
-               reg = <2>;
+       /* External ADC(+) */
+       channel@5 {
+               reg = <5>;
                zephyr,gain = "ADC_GAIN_1";
                zephyr,reference = "ADC_REF_EXTERNAL0";
                zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
-               zephyr,input-positive = <4>;
-               zephyr,input-negative = <5>;
+               zephyr,input-positive = <5>;
                zephyr,resolution = <12>;
        };

```

```
*** Booting Zephyr OS build v3.3.0-rc2-204-g86548f258c98 ***
ADC reading:
- adc@40038000, channel 5: 64 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 1716 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 1720 = 0 mV
ADC reading:
- adc@40038000, channel 5: 60 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 62 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 61 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 61 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 1726 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 1724 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 62 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 62 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 63 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 43 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 1754 = 0 mV
ADC reading:
- adc@40038000, channel 5: 39 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 67 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 146 = 0 mV
- adc@40038000, channel 15: 1717 = 0 mV
ADC reading:
- adc@40038000, channel 5: 1841 = 0 mV
- adc@40038000, channel 15: 1723 = 0 mV
ADC reading:                                                                       
- adc@40038000, channel 5: 3734 = 0 mV                                             
- adc@40038000, channel 15: 2047 = 0 mV
```

teburd commented:
@pdgendt please take a look

nandojve commented:
Hi @pdgendt ,

Any new driver should go after #55655.

pdgendt commented:
> Any new driver should go after #55655.

Hi @nandojve, I will update the PR once merged, for the test on `sam4s_xplained`, I think it's a clocking setting that should be tweaked.

Could you test again with the `prescaler` property set to the maximum `255`? I've done the same on my custom board and the internal temperature sensor is stable.

I don't know what sensible values would be for the timings on the `sam4s_xplained` board, any suggestions?

nandojve commented:
Hi @pdgendt ,

Thank you for look on it. I'll have time to look on HW at weekend. Right now my reference have been the example inside [Atmel ASF](https://ww1.microchip.com/downloads/en/DeviceDoc/asf-standalone-archive-3.52.0.113.zip) at ASF->SAM->Drivers->ADC.

pdgendt commented:
> Thank you for look on it. I'll have time to look on HW at weekend. Right now my reference have been the example inside [Atmel ASF](https://ww1.microchip.com/downloads/en/DeviceDoc/asf-standalone-archive-3.52.0.113.zip) at ASF->SAM->Drivers->ADC.

Great, thanks.

I used the example as a reference for my implementation, setting the `prescaler` did yield stable readings, so I guess the driver is functional.

Maybe the bindings could improve, I currently set these values as they would be set in the `MR` register. I could, for example, change the `prescaler` binding to `frequency` binding and calculate the value using your implementation for the peripheral clock.

nandojve commented on boards/arm/sam4s_xplained/sam4s_xplained.dts:
```diff
@@ -176,6 +176,18 @@
+	pinctrl-names = "default";
+
+	prescaler = <2>;
+	startup-time = <1>;
```
startup-time = <4>;

nandojve commented on boards/arm/sam4s_xplained/sam4s_xplained.dts:
```diff
@@ -176,6 +176,18 @@
+	prescaler = <2>;
+	startup-time = <1>;
+	settling-time = <0>;
+	tracking-time = <2>;
```
tracking-time = <1>;

nandojve commented on drivers/adc/adc_sam.c:
```diff
@@ -0,0 +1,381 @@
+	adc->ADC_MR = ADC_MR_PRESCAL(cfg->prescaler)
+		    | ADC_MR_STARTUP(cfg->startup_time)
+		    | ADC_MR_SETTLING(cfg->settling_time)
+		    | ADC_MR_TRACKTIM(cfg->tracking_time)
```
```diff
+                   | ADC_MR_TRANSFER(2)
```

nandojve reviewed (changes requested):
Hi @pdgendt ,

I still not get good results. I don't know why but potentiometer oscillates which not happen with Atmel sample code, maybe a concurrency problem?

pdgendt commented on drivers/adc/adc_sam.c:
```diff
@@ -0,0 +1,381 @@
+			LOG_ERR("ADC 1/2x gain only allowed for differential channel");
+			return -EINVAL;
+		}
+		/* NOP */
```
0.5 gain is the default value after resetting, which is done before the switch/case with:

```c
	adc->ADC_CGR &= ~(ADC_CGR_GAIN0_Msk << (channel_id * 2U));
```

pdgendt commented on dts/bindings/adc/atmel,sam-adc.yaml:
```diff
@@ -0,0 +1,86 @@
+    description: peripheral ID
+    required: true
+
+  prescaler:
+    type: int
+    required: true
+    description: CPU clock prescaler applied to get the ADC clock.
```
The prescaler applies to all channels, can we use the `channel_cfg->acquisition_time` that applies to a single channel?

nandojve replied:
Let's do this in another PR.

pdgendt commented:
Hi @nandojve,

Did you test using `prescaler = <10>`? This would yield an ADC frequency of 5,4 MHz (`120 / 2 / (prescaler + 1`). Which is higher than the allowed maximum of 1 MHz.

The minimum value is `prescaler = <59>` if I applied the correct formula (`120 / 2 / (59 + 1) = 1`).

pdgendt commented on drivers/adc/adc_sam.c:
```diff
@@ -0,0 +1,381 @@
+	adc->ADC_RCR = 0U;
+	adc->ADC_RNCR = 0U;
+
+	/* Set prescaler, timings and allow different analog settings for each channel */
+	adc->ADC_MR = ADC_MR_PRESCAL(cfg->prescaler)
+		    | ADC_MR_STARTUP(cfg->startup_time)
+		    | ADC_MR_SETTLING(cfg->settling_time)
+		    | ADC_MR_TRACKTIM(cfg->tracking_time)
+		    | ADC_MR_ANACH_ALLOWED;
```
See comment https://github.com/zephyrproject-rtos/zephyr/pull/51885#discussion_r1141749364

nandojve reviewed (approved):
LGTM!
